x/protocolpool/keeper/keeper.go

```diff
@@ -0,0 +1,226 @@
+}
+
+// GetCommunityPool gets the community pool balance.
+func (k Keeper) GetCommunityPool(ctx sdk.Context) (sdk.Coins, error) {
+	moduleAccount := k.authKeeper.GetModuleAccount(ctx, types.ModuleName)
+	if moduleAccount == nil {
+		return nil, errorsmod.Wrapf(sdkerrors.ErrUnknownAddress, "module account %s does not exist", types.ModuleName)
+	}
+	return k.bankKeeper.GetAllBalances(ctx, moduleAccount.GetAddress()), nil
+}
+
+// DistributeFunds sets the amount to be distributed among recipients.
+// Get all valid continuous funds:
+// - for each continuous fund, check if expired and remove if so
+// - for each continuous fund, distribute funds according to percentage
+// - distribute remaining funds to the community pool
+//
+// This function is run at the BeginBlocker method and therefore must be very safe.
+func (k Keeper) DistributeFunds(ctx sdk.Context) error {
+	// Get current balance of the intermediary module account
+	moduleAccount := k.authKeeper.GetModuleAccount(ctx, types.ProtocolPoolEscrowAccount)
+	if moduleAccount == nil {
+		return errorsmod.Wrapf(sdkerrors.ErrUnknownAddress, "module account %s does not exist", types.ProtocolPoolEscrowAccount)
+	}
+	params, err := k.Params.Get(ctx)
+	if err != nil {
+		return err
+	}
+
+	// only take into account the balances of denoms whitelisted in EnabledDistributionDenoms
+	amountToDistribute := sdk.NewCoins()
+	for _, denom := range params.EnabledDistributionDenoms {
+		bal := k.bankKeeper.GetBalance(ctx, moduleAccount.GetAddress(), denom)
+		amountToDistribute = append(amountToDistribute, bal)
+	}
+
+	// if the balance is zero, return early
+	if amountToDistribute.IsZero() {
+		return nil
+	}
+
+	remainingCoins := sdk.NewCoins(amountToDistribute...)
+	iter, err := k.ContinuousFunds.Iterate(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("failed to create iterator for continuous funds: %w", err)
+	}
+
+	kvalues, err := iter.KeyValues()
+	if err != nil {
+		return fmt.Errorf("failed to iterate continuous funds: %w", err)
+	}
+
+	blockTime := ctx.BlockTime()
+	anyNegative := false
+	for _, kv := range kvalues {
+		recipient := kv.Key
+		fund := kv.Value
+
+		// remove newly expired funds
+		if fund.Expiry != nil && fund.Expiry.Before(blockTime) {
+			err := k.ContinuousFunds.Remove(ctx, recipient)
+			if err != nil {
+				return fmt.Errorf("failed to remove fund for %s from ContinuousFunds: %w", recipient, err)
+			}
+			continue
+		}
+
+		amountToStream := PercentageCoinMul(fund.Percentage, amountToDistribute)
+		remainingCoins, anyNegative = remainingCoins.SafeSub(amountToStream...)
+		if anyNegative {
+			return fmt.Errorf("negative funds for distribution from ContinuousFunds: %v", remainingCoins)
+		}
+
+		err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ProtocolPoolEscrowAccount, recipient, amountToStream)
+		if err != nil {
+			return fmt.Errorf("failed to distribute fund for %s from ContinuousFunds: %w", recipient, err)
+		}
```

Comment:
we should check that the recipient is not a blocked address in bank via the error returned

maybe also deny the creation of continuous funds that are blocked by bank